File: backend/duty.py
"""جدول الإجمالي اليومي — تصنيف كل ضابط في خانة واحدة بناءً على تشغيله."""
from .constants import SHIFTS, LEAVE_BUCKET
from .leaves import leave_on
from .people import officers_on


def summarise(data, day):
    """جدول الإجمالي أسفل يومية الضباط — كل ضابط في خانة واحدة بس."""
    services = {s["id"]: s for s in data["services"]}
    duties = data["duties"].get(day, {})
    officers = officers_on(data, day)

    s = {
        "أصل القوة": len(officers),
        "خارجية": {"صباحية": 0, "ليلية": 0, "بحث": 0},
        "داخلية": {"صباحية": 0, "ليلية": 0},
        "طبية": {"موجود": 0, "راحة": 0},
        "خوارج": {k: 0 for k in ("تقصيرة", "راحة", "طارئة", "غياب", "مرضي", "فرقة", "انتداب")},
        "حراسات": 0,
        "صافي": 0,
    }
    net_names, rows = [], []

    for o in officers:
        d = duties.get(o["id"], {})
        kinds = [(services.get(i.get("service_id"), {}), i.get("shift", "صباحية"))
                 for i in d.get("items", [])]
        lv = leave_on(data, o["id"], day)
        medical = any(sv.get("kind") == "طبية" for sv, _ in kinds)

        # الترتيب هنا هو نفس ترتيب الأولوية في اليومية الورقية
        if medical:
            bucket = ("طبية", "راحة" if lv else "موجود")
        elif lv:
            bucket = ("خوارج", LEAVE_BUCKET.get(lv["type"], "راحة"))
        elif d.get("status") in ("انتداب", "غياب"):
            bucket = ("خوارج", d["status"])
        elif d.get("taqseera"):
            bucket = ("خوارج", "تقصيرة")
        elif any(sv.get("kind") == "حراسات" for sv, _ in kinds):
            bucket = ("حراسات", None)
        elif any(sv.get("kind") == "داخلية" for sv, _ in kinds):
            sh = next(sh for sv, sh in kinds if sv.get("kind") == "داخلية")
            bucket = ("داخلية", sh if sh in SHIFTS else "صباحية")
        elif any(sv.get("kind") == "بحث" for sv, _ in kinds):
            bucket = ("خارجية", "بحث")
        elif any(sv.get("kind") == "خارجية" for sv, _ in kinds):
            sh = next(sh for sv, sh in kinds if sv.get("kind") == "خارجية")
            bucket = ("خارجية", sh if sh in SHIFTS else "صباحية")
        else:
            bucket = ("صافي", None)

        grp, sub = bucket
        if sub is None:
            s[grp] += 1
            if grp == "صافي":
                net_names.append(f'{o.get("role","")}/ {o.get("name","")}')
        else:
            s[grp][sub] += 1

        rows.append({
            "id": o["id"], "name": o.get("name", ""), "role": o.get("role", ""),
            "group": grp, "bucket": sub,
            "services": [{"id": sv.get("id"), "name": sv.get("name"), "kind": sv.get("kind"),
                          "shift": sh} for sv, sh in kinds if sv],
            "taqseera": bool(d.get("taqseera")),
            "leave": ({"type": lv["type"], "start": lv["start"], "end": lv["end"],
                      "return_date": lv["return_date"]} if lv else None),
            "note": d.get("note", ""),
            # كان بالقوة يوم كذا لكنه خرج بعد كده — للتوضيح في اليوميات القديمة
            "later_left": o.get("leave_date", "") or None,
        })

    counted = (sum(s["خارجية"].values()) + sum(s["داخلية"].values())
               + sum(s["طبية"].values()) + sum(s["خوارج"].values())
               + s["حراسات"] + s["صافي"])
    s["net_names"] = net_names
    s["balanced"] = counted == s["أصل القوة"]
    s["counted"] = counted
    return {"date": day, "summary": s, "rows": rows}

File: static/js/duty.js
/* يومية تشغيل الضباط — جدول الإجمالي وحالة الضابط.
   التكليف بخدمة بقى في اليومية التفصيلية (نفس السجل)، والصفحة دي بتعدّل
   حالة الضابط بس: تقصيرة / انتداب / غياب / مرضي / فرقة / طارئة / ملاحظة. */
let DUTY = null;

function renderSummary(s) {
  const cell = v => `<td>${v}</td>`;
  const x = s["خارجية"], dl = s["داخلية"], tb = s["طبية"], kh = s["خوارج"];
  $("#balanceTag").innerHTML = s.balanced
    ? `<span class="chip on">متوازن ${s.counted}/${s["أصل القوة"]}</span>`
    : `<span class="chip rest">غير متوازن ${s.counted}/${s["أصل القوة"]}</span>`;
  $("#dutySummary").innerHTML = `<div class="table-scroll"><table class="table sum">
   <thead>
    <tr><th rowspan="2">أصل القوة</th><th colspan="3">الخدمات الخارجية</th><th colspan="2">الخدمات الداخلية</th>
        <th colspan="2">الخدمات الطبية</th><th colspan="7">الخوارج</th>
        <th rowspan="2">الحراسات المشددة</th><th rowspan="2">الصافي (${s["صافي"]})</th></tr>
    <tr><th>صباحية</th><th>ليلية</th><th>بحث</th><th>صباحية</th><th>ليلية</th>
        <th>موجود</th><th>راحة</th>
        <th>تقصيرة</th><th>راحة</th><th>طارئة</th><th>غياب</th><th>مرضي</th><th>فرقة</th><th>انتداب</th></tr>
   </thead>
   <tbody><tr class="sumrow">
     <td><b>${s["أصل القوة"]}</b></td>
     ${cell(x["صباحية"])}${cell(x["ليلية"])}${cell(x["بحث"])}
     ${cell(dl["صباحية"])}${cell(dl["ليلية"])}
     ${cell(tb["موجود"])}${cell(tb["راحة"])}
     ${cell(kh["تقصيرة"])}${cell(kh["راحة"])}${cell(kh["طارئة"])}${cell(kh["غياب"])}${cell(kh["مرضي"])}${cell(kh["فرقة"])}${cell(kh["انتداب"])}
     ${cell(s["حراسات"])}
     <td class="wrap">${s.net_names.map(esc).join("<br>") || "—"}</td>
   </tr></tbody></table></div>`;
}

/* أقسام جدول الوورد الثلاثة — القسم تابع لوضع الضابط التنظيمي مش لتشغيله */
function sectionRows(rows) {
  const order = META.officer_sections || ["القوة", "الحراسات المشددة", "الخوارج"];
  const out = [];
  for (const section of order) {
    const mine = rows.filter(r => (r.section || order[0]) === section);
    if (!mine.length) continue;
    if (section !== order[0]) {
      out.push(`<tr class="section-row"><td colspan="6">${esc(section)}</td></tr>`);
    }
    out.push(...mine.map(officerRow));
  }
  // أي قسم مش في القايمة يتعرض في الآخر بدل ما يختفي
  out.push(...rows.filter(r => !order.includes(r.section || order[0])).map(officerRow));
  return out;
}

function officerRow(r) {
  const svc = r.services.length
    ? r.services.map(s => `<span class="chip ${KIND_CLS[s.kind] || "w"}">${esc(s.name)}<i>${esc(s.shift)}</i></span>`).join(" ")
    : `<span class="muted">—</span>`;
  const grp = r.leave ? `<span class="chip rest">${esc(r.leave.type)}</span><div class="sub">حتى ${fmt(r.leave.end)}</div>`
    : r.group === "صافي" ? `<span class="chip on">صافي</span>`
    : `<span class="chip ${KIND_CLS[r.group] || "done"}">${esc(r.group)}${r.bucket ? " · " + esc(r.bucket) : ""}</span>`;
  const gone = r.later_left ? `<span class="chip done" title="خرج من القوة يوم ${fmt(r.later_left)}">خرج ${fmt(r.later_left)}</span>` : "";
  const search = r.search_attached ? ` <span class="chip soon" title="تشغيل من إدارة البحث">بحث</span>` : "";
  return `<tr class="${r.later_left ? "was" : ""}">
    <td class="name">${esc(r.name)}<div class="sub">${esc(r.role)}${search} ${gone}</div></td>
    <td class="wrap">${esc(r.post) || "<span class='muted'>—</span>"}</td>
    <td>${svc}${r.taqseera ? ' <span class="chip taq">تقصيرة</span>' : ""}</td>
    <td>${grp}</td>
    <td class="wrap">${esc(r.note) || "<span class='muted'>—</span>"}</td>
    <td><button class="mini" data-action="openAssign" data-id="${esc(r.id)}">الحالة</button></td>
  </tr>`;
}

function render() {
  if (!DUTY) { $("#dutyBoard").innerHTML = `<div class="empty">جارٍ التحميل...</div>`; return }
  renderSummary(DUTY.summary);
  const gone = DUTY.rows.filter(r => r.later_left).length;
  $("#dutyBoard").innerHTML = tableBlock(
    ["الضابط", "العمل المسند إليه", "الخدمات", "الخانة في الإجمالي", "نص التشغيل", "الإجراء"],
    sectionRows(DUTY.rows),
    `قوة اليوم: ${DUTY.rows.length} ضابط${gone ? ` — منهم ${gone} خرجوا من القوة بعد كده` : ""}`);
}

ACTIONS.openAssign = id => {
  const row = DUTY.rows.find(r => r.id === id); if (!row) return;
  $("#assignPerson").value = id;
  $("#assignTitle").textContent = `حالة: ${row.name}`;
  $("#assignHint").textContent =
    `يوم ${dayName(DUTY.date)} ${fmt(DUTY.date)} — التكليف بالخدمات من اليومية التفصيلية.`
    + (row.leave ? ` تنبيه: الضابط في ${row.leave.type} لحد ${fmt(row.leave.end)}.` : "");
  fillSelect($("#assignStatus"),
    [["", "— بدون —"], ...(META.officer_statuses || []).map(x => [x, x])]);
  $("#assignStatus").value = row.group === "خوارج" && !row.leave ? (row.bucket || "") : "";
  $("#assignTaq").checked = !!row.taqseera;
  $("#assignNote").value = row.note || "";
  openModal("assignModal");
};

$("#assignForm").onsubmit = async e => {
  e.preventDefault();
  const out = await api(`/api/duty/${DUTY.date}/${encodeURIComponent($("#assignPerson").value)}`,
    jsonReq("PUT", {taqseera: $("#assignTaq").checked,
      status: $("#assignStatus").value, note: $("#assignNote").value}));
  if (!out) return;
  DUTY = out; closeModal("assignModal"); showToast("تم حفظ الحالة"); render();
};

async function loadDay(day) {
  const d = await api(`/api/duty/${day}`); if (!d) return;
  DUTY = d; render();
}
const shiftDay = n => { const d = addDays($("#dutyDate").value || curDate(), n); $("#dutyDate").value = d; loadDay(d) };
$("#dayPrev").onclick = () => shiftDay(-1);
$("#dayNext").onclick = () => shiftDay(1);
$("#dayToday").onclick = () => { $("#dutyDate").value = curDate(); loadDay(curDate()) };
$("#dutyDate").onchange = () => loadDay($("#dutyDate").value);

async function load() {
  const d = await bootstrap();
  if (!d) return;
  const days = d.days || [];
  $("#dutyDate").value = days.includes(curDate()) ? curDate() : (days[days.length - 1] || curDate());
  loadDay($("#dutyDate").value);
}
load();
